### ISSUE TITLE
AttributeError: 'NoneType' object has no attribute 'create_token_response'

### ISSUE THREAD

**Issue by jer-tx** [tilgovi/pyramid-oauthlib#18]
Given the lack of documentation, I'm not 100% sure here. I think I have a basic implementation but I'm running into this error. Within my config, I've added:

```
pyramid_oauthlib.includeme(config)
config.add_response_type(AuthorizationCodeGrant,
                             name='code',
                             request_validator=validator)
config.add_grant_type(AuthorizationCodeGrant,
                          name='authorization_code',
                          request_validator=validator)

config.add_token_type(BearerToken,
                          request_validator=validator)
```

And then on my handler, just to get a proof of concept/see it work, I've just added:

```
@view_config(route_name='oauth', request_method="POST", renderer="json")
def oauth(request):
    return request.create_token_response()
```
Calling that endpoint with postman, I get this traceback

```
  File "C:\Users\Jeremy\PycharmProjects\comparebench\comparebench\view_handlers\user_handler.py", line 25, in oauth
    return request.create_token_response()
  File "C:\Users\Jeremy\AppData\Local\Programs\Python\Python38-32\lib\site-packages\pyramid_oauthlib\__init__.py", line 222, in <lambda>
    server.create_token_response(request, credentials=credentials)
  File "C:\Users\Jeremy\AppData\Local\Programs\Python\Python38-32\lib\site-packages\oauthlib\oauth2\rfc6749\endpoints\base.py", line 64, in wrapper
    return f(endpoint, uri, *args, **kwargs)
  File "C:\Users\Jeremy\AppData\Local\Programs\Python\Python38-32\lib\site-packages\pyramid_oauthlib\__init__.py", line 85, in create_token_response
    return handler.create_token_response(request, token)
AttributeError: 'NoneType' object has no attribute 'create_token_response'
```

**Comment by tilgovi** [comment]
If I remember correctly, though it's been a while, I set this library up so that there's a quick setup in which you don't name things and then act as defaults.

The error you're getting is caused by there being no matching grant type. Either pass a `grant_type=authorization_code` as a request parameter, or removing the `name='authorization_code'` when you call `config.add_grant_type` to register it as the default grant type.

You can also register it with or without a name. If you can think of a good API for registering named {grant types, response codes, token types} that are also the default, please suggest it!

**Comment by jer-tx** [comment]
That seemed to get me further, but getting a new error:

```
File "C:\Users\Jeremy\PycharmProjects\comparebench\comparebench\view_handlers\user_handler.py", line 25, in oauth
    return request.create_token_response()
  File "C:\Users\Jeremy\AppData\Local\Programs\Python\Python38-32\lib\site-packages\pyramid_oauthlib\__init__.py", line 222, in <lambda>
    server.create_token_response(request, credentials=credentials)
  File "C:\Users\Jeremy\AppData\Local\Programs\Python\Python38-32\lib\site-packages\oauthlib\oauth2\rfc6749\endpoints\base.py", line 64, in wrapper
    return f(endpoint, uri, *args, **kwargs)
  File "C:\Users\Jeremy\AppData\Local\Programs\Python\Python38-32\lib\site-packages\pyramid_oauthlib\__init__.py", line 85, in create_token_response
    return handler.create_token_response(request, token)
  File "C:\Users\Jeremy\AppData\Local\Programs\Python\Python38-32\lib\site-packages\oauthlib\oauth2\rfc6749\grant_types\authorization_code.py", line 242, in create_token_response
    self.validate_token_request(request)
  File "C:\Users\Jeremy\AppData\Local\Programs\Python\Python38-32\lib\site-packages\oauthlib\oauth2\rfc6749\grant_types\authorization_code.py", line 374, in validate_token_request
    raise errors.InvalidRequestError(
  File "C:\Users\Jeremy\AppData\Local\Programs\Python\Python38-32\lib\site-packages\oauthlib\oauth2\rfc6749\errors.py", line 61, in __init__
    self.response_mode = request.response_mode
AttributeError: 'Request' object has no attribute 'response_mode'
```

**Comment by tilgovi** [comment]
What version of oauthlib?

**Comment by jer-tx** [comment]
2.1.0

**Comment by tilgovi** [comment]
It looks like I missed `response_mode` in the parameter list: https://github.com/tilgovi/pyramid-oauthlib/blob/139985514300c93206082dde940bca968f533bae/pyramid_oauthlib/__init__.py#L12

You could try calling `config.add_oauth_param('response_mode')` as a workaround.

**Comment by jer-tx** [comment]
New error:

```
  File "C:\Users\Jeremy\PycharmProjects\comparebench\comparebench\view_handlers\user_handler.py", line 25, in oauth
    return request.create_token_response()
  File "C:\Users\Jeremy\AppData\Local\Programs\Python\Python38-32\lib\site-packages\pyramid_oauthlib\__init__.py", line 222, in <lambda>
    server.create_token_response(request, credentials=credentials)
  File "C:\Users\Jeremy\AppData\Local\Programs\Python\Python38-32\lib\site-packages\oauthlib\oauth2\rfc6749\endpoints\base.py", line 64, in wrapper
    return f(endpoint, uri, *args, **kwargs)
  File "C:\Users\Jeremy\AppData\Local\Programs\Python\Python38-32\lib\site-packages\pyramid_oauthlib\__init__.py", line 85, in create_token_response
    return handler.create_token_response(request, token)
  File "C:\Users\Jeremy\AppData\Local\Programs\Python\Python38-32\lib\site-packages\oauthlib\oauth2\rfc6749\grant_types\authorization_code.py", line 242, in create_token_response
    self.validate_token_request(request)
  File "C:\Users\Jeremy\AppData\Local\Programs\Python\Python38-32\lib\site-packages\oauthlib\oauth2\rfc6749\grant_types\authorization_code.py", line 398, in validate_token_request
    if not hasattr(request.client, 'client_id'):
AttributeError: 'Request' object has no attribute 'client'
```

Tried adding it to config like the response mode but, seems it needs a Client object. Not sure where in the flow of all this that would be added

**Comment by jer-tx** [comment]
Got it, at least for creating a token. 

I added `request.client = Client(client_id=request.client_id)` to the `create_token_response` method of `Server()`

That seemed to get the ball rolling here :)

**Comment by jer-tx** [comment]
Created https://github.com/tilgovi/pyramid-oauthlib/pull/19 for these changes

**Comment by tilgovi** [comment]
The context of that attribute error is this code in oauthlib:

``` python
        if not hasattr(request.client, 'client_id'):
            raise NotImplementedError('Authenticate client must set the '
                                      'request.client.client_id attribute '
                                      'in authenticate_client.')
```

So the expectation is that the request validator should initialize `request.client`. Implement the `authenticate_client_id(self, client_id, request)` method on your request validator and set `request.client` there.

**Comment by jer-tx** [comment]
Ah, interesting. I've switched to `ResourceOwnerPasswordCredentialsGrant` but the solution is similar. Need to set `request.client` on `authenticate_client`. I'll remove those from my PR

**Comment by tilgovi** [comment]
Closing after merging #19. Please open new issues for any other bugs you encounter, or any questions you have, or any suggestions you have!

**Comment by tilgovi** [comment]
I published v0.4.2.
